lib/ansible/modules/network/panos/panos_security_policy.py

```diff
@@ -0,0 +1,518 @@
+        if 'spyware' in kwargs:
+            security_rule.spyware = kwargs['spyware']
+        if 'url_filtering' in kwargs:
+            security_rule.url_filtering = kwargs['url_filtering']
+        if 'file_blocking' in kwargs:
+            security_rule.file_blocking = kwargs['file_blocking']
+        if 'data_filtering' in kwargs:
+            security_rule.data_filtering = kwargs['data_filtering']
+        if 'wildfire_analysis' in kwargs:
+            security_rule.wildfire_analysis = kwargs['wildfire_analysis']
+
+    return security_rule
+
+
+def add_security_rule(device, sec_rule):
+    if isinstance(device, pandevice.firewall.Firewall):
+        rule_base = pandevice.policies.Rulebase.refreshall(device)
+    elif isinstance(device, pandevice.panorama.Panorama):
+        # look for only pre-rulebase ATM
+        rule_base = pandevice.policies.PreRulebase.refreshall(device)
+
+    if rule_base:
+        rule_base = rule_base[0]
+
+        rule_base.add(sec_rule)
+        sec_rule.create()
+
+        return True
+    else:
+        return False
+
+
+def _commit(device, device_group=None):
+    """
+    :param device: either firewall or panorama
+    :param device_group: panorama device group or if none then 'all'
+    :return: True if successful
+    """
+    result = device.commit(sync=True)
+
+    if isinstance(device, pandevice.panorama.Panorama):
+        result = device.commit_all(sync=True, sync_all=True, devicegroup=device_group)
+
+    return result
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(no_log=True),
+        username=dict(default='admin'),
+        api_key=dict(no_log=True),
+        rule_name=dict(required=True),
+        description=dict(default=''),
+        tag=dict(),
+        to_zone=dict(type='list', default=['any']),
+        from_zone=dict(type='list', default=['any']),
+        source=dict(type='list', default=["any"]),
+        source_user=dict(type='list', default=['any']),
+        destination=dict(type='list', default=["any"]),
+        category=dict(type='list', default=['any']),
+        application=dict(type='list', default=['any']),
+        service=dict(type='list', default=['application-default']),
+        hip_profiles=dict(type='list', default=['any']),
+        group_profile=dict(),
+        antivirus=dict(),
+        vulnerability=dict(),
+        spyware=dict(),
+        url_filtering=dict(),
+        file_blocking=dict(),
+        data_filtering=dict(),
+        wildfire_analysis=dict(),
+        log_start=dict(type='bool', default=False),
+        log_end=dict(type='bool', default=True),
+        rule_type=dict(default='universal'),
+        action=dict(default='allow'),
+        devicegroup=dict(),
+        commit=dict(type='bool', default=True)
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False,
+                           required_one_of=[['api_key', 'password']])
```

Comment:
It is not really mutually exclusive as code logic will try multiple authentication methods. It is more of 'unnecessary' case.